### PR TITLE
Image support

### PR DESCRIPTION
**Pull request by tal5**
## Additions

- `ImageTag` object type, represents an image.
- `BinaryTag#isValidHex` - returns whether a string is a valid binary hex.
- `ColorTag#asARGB` - returns a ARGB `int` of the color.
- `ColorTag#fromARGB` - gets a `ColorTag` from an ARGB int.
- `draw` command - draws a pixel/shapes onto an image.
- `drawimage` command - draws an image onto another image.
- `image` command - handles loading, saving, and unloading of images by id.
- `FileRead/WriteCommand#getFileIfSafe` - takes in a path string and returns a file, or `null` if it isn't accessible.
- `BinaryTag.to_image` - tries converting the `BinaryTag` into an `ImageTag`.

## Changes

- Fixed a minor typo in `BinaryTag`'s object type meta.
- `BinaryTag#valueOf/matches` now use the new `BinaryTag#isValidHex`.
- `FileRead/WriteCommand` now use the new `getFileIfSafe` methods.
- Fixed a minor typo in the `YAML` command's meta.

## Notes

- In the `image` command's meta I had to specify the path as `plugins/Denizen`, as there isn't really any better non impl-specific way to explain that currently.
- There's sill an optimization to be made with the way the `adjust` command clones objects, see [Discord](https://discord.com/channels/315163488085475337/1011496047811506227/1172926529647415327).

### CODE REVIEW

**Comment by mcmonkey4eva** [comment]
Marking this a draft til it's fully implemented (ie has the id tracker and all rather than hardcopying endlessly)

File: src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java [hunk 1]
```diff
@@ -52,55 +53,63 @@ public FileReadCommand() {
...
+        };
+        if (scriptEntry.shouldWaitFor()) {
+            DenizenCore.runAsync(runme);
+        }
+        else {
+            runme.run();
+        }
+    }
+
+    public static File canRead(String path, ScriptEntry scriptEntry) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java:81]: This is more of a `tryReadSafe` than a `canRead`

File: src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java [hunk 2]
```diff
@@ -52,55 +53,63 @@ public FileReadCommand() {
...
         if (!CoreConfiguration.allowFileRead) {
-            Debug.echoError(scriptEntry, "FileRead disabled in Denizen/config.yml (refer to command documentation).");
+            Debug.echoError(scriptEntry, "FileRead disabled in configuration (refer to command documentation).");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java:83]: `in configuration`, is (A) not English and (B) not really clear - what configuration? If altering this to be compatible with non-bukkit impl (ie clientizen) there should probably be some CoreImpl string that fills this spot with something coherent local to the impl

File: src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        // @returns ElementTag
+        // @description
+        // Returns the image's type ("png", "jpg", "webp", etc.).
+        // @example
+        // Checks if an image is a png.
+        // - if <[image].type> == png:
+        //   - narrate "It's a png!"
+        // -->
+        tagProcessor.registerStaticTag(ElementTag.class, "type", (attribute, object) -> {
+            return new ElementTag(object.imageType, true);
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java:211]: Why is this tracked on the image? This is a file property not an image property

Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java:211]: It's needed to write the image later (see `getRawBytes`, `image` command, etc.)

File: src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    public void markChanged() {
+        hexEncodedBytesCache = null;
+    }
+
+    @Override
+    public String identify() {
+        if (hexEncodedBytesCache == null) {
+            hexEncodedBytesCache = BinaryTag.hexEncode(getRawBytes(), false);
+        }
+        return "image@" + hexEncodedBytesCache;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java:330]: Probably should store `image@` in the cache so it's not copying huge strings in memory here

File: src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java (hunk 3 above)
Comment (mcmonkey4eva) [reply src/main/java/com/denizenscript/denizencore/objects/core/ImageTag.java:211]: Should be inputs in those locations then?

**Comment by tal5** [comment]
Should all be fixed now - I've marked it as ready for review

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import com.denizenscript.denizencore.scripts.commands.generator.ArgName;
+import com.denizenscript.denizencore.scripts.commands.generator.ArgPrefixed;
+
+import java.awt.*;
+
+public class DrawCommand extends AbstractCommand {
+
+    // <--[command]
+    // @Name Draw
+    // @Syntax draw [id:<id>] [PIXEL]/[RECTANGLE/OVAL [width:<width>] [height:<height>] (filled)] [x:<x>] [y:<y>] [color:<color>]
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java:17]: shouldn't be caps here

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    // @Group image
+    //
+    // @Description
+    // Draws onto an image, potentially with certain predetermined shapes.
+    //
+    // "id:" is the id of the image to draw on, see <@link command Image>.
+    // "color:" is a <@link ObjectType ColorTag> of the color to draw in.
+    // "x:" and "y:" are the position that should be drawn on, see <@link language Image positions>.
+    //
+    // If you are drawing a shape (not a pixel), you can also use:
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java:30]: `must` not `can`

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawImageCommand.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import com.denizenscript.denizencore.scripts.commands.generator.ArgName;
+import com.denizenscript.denizencore.scripts.commands.generator.ArgPrefixed;
+
+import java.awt.*;
+
+public class DrawImageCommand extends AbstractCommand {
+
+    // <--[command]
+    // @Name DrawImage
+    // @Syntax drawimage [id:<id>] [draw:<image>] [x:<x>] [y:<y>] (width:<width>) (height:<height>)
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawImageCommand.java:15]: This command looks like it'd be in cleanly inside of `draw`?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    //
+    // @Tags
+    // None
+    //
+    // @Usage
+    // Use to load an image from a file into memory under the id "image".
+    // - ~image id:image load path:data/image.png
+    //
+    // @Usage
+    // Use to load an image object from a definition, draw on it, then store it back into the definition.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java:55]: This example should be updated per image-identified-by-id change

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                if (toLoad == null && path == null) {
+                    throw new InvalidArgumentsRuntimeException("Must specify a path or image to load.");
+                }
+                if (loadedImages.containsKey(idLower)) {
+                    Debug.echoError(scriptEntry, "Id '" + id + "' is already in use.");
+                    scriptEntry.setFinished(true);
+                    return;
+                }
+                if (toLoad != null) {
+                    ImageTag toLoadCopy = toLoad.duplicate();
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java:104]: This block needs an update for images-identified-by-id change

File: src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java [hunk 10]
```diff
@@ -52,55 +53,63 @@ public FileReadCommand() {
...
             if (!file.exists()) {
                 Debug.echoError(scriptEntry, "File read failed, file does not exist!");
                 scriptEntry.setFinished(true);
-                return;
+                return null;
             }
         }
-        catch (Exception ex) {
-            Debug.echoError(ex);
+        catch (IOException e) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java:108]: Why was this changed?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java (hunk 5 above)
Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java]: Corrected

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java (hunk 6 above)
Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawCommand.java]: Fixed

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java (hunk 8 above)
Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java:55]: Repaired

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java (hunk 9 above)
Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/core/ImageCommand.java]: Mended

File: src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java (hunk 10 above)
Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/file/FileReadCommand.java]: Did Intellij...? either way, reverted

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawImageCommand.java (hunk 7 above)
Comment (tal5) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/core/DrawImageCommand.java]: Merged, lmk if the syntax looks good
